keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,579 @@
+	// +optional
+	FailedJobsHistoryLimit *int32
+}
+
+// ConcurrencyPolicy describes how the job will be handled.
+// Only one of the following concurrent policies may be specified.
+// If none of the following policies is specified, the default one
+// is AllowConcurrent.
+type ConcurrencyPolicy string
+
+const (
+	// AllowConcurrent allows CronJobs to run concurrently.
+	AllowConcurrent ConcurrencyPolicy = "Allow"
+
+	// ForbidConcurrent forbids concurrent runs, skipping next run if previous
+	// hasn't finished yet.
+	ForbidConcurrent ConcurrencyPolicy = "Forbid"
+
+	// ReplaceConcurrent cancels currently running job and replaces it with a new one.
+	ReplaceConcurrent ConcurrencyPolicy = "Replace"
+)
+
+// CronJobConditionType for CronJob conditions
+type CronJobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// CronJobSettled means the cron job controller has
+	CronJobSettled JobConditionType = "Settled"
+)
+
+// CronJobCondition describes a condition state
+type CronJobCondition struct {
+	// Type of job condition, Settled
+	Type CronJobConditionType
+	// Status of the condition, one of True, False, Unknown.
+	Status api.ConditionStatus
+	// Last time the condition was checked.
+	// +optional
+	LastProbeTime metav1.Time
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// (brief) reason for the condition's last transition.
+	// +optional
+	Reason string
+	// Human readable message indicating details about last transition.
+	// +optional
+	Message string
+}
+
+// CronJobStatus represents the current state of a cron job.
+type CronJobStatus struct {
+	// A list of pointers to currently running jobs.
+	// +optional
+	Active []api.ObjectReference
+
+	// Information when was the last time the job was successfully scheduled.
+	// +optional
+	LastScheduleTime *metav1.Time
+
+	// Information for when was the next time the job will be scheduled.
+    // May not be accurate in some cases for Forbid concurrency policy
+	// +optional
+	NextScheduleTime *metav1.Time
+
+	// Information when was the last time the job successfully completed.
+	// +optional
+	LastSuccessfulTime *metav1.Time
+
+	// Counter for tracking successful Job runs
+	// +optional
+	SuccessfulRuns int64
+
+	// Counter for tracking failed Job runs
+	// +optional
+    FailedlRuns int64
+
+	// Counter for tracking failed Job runs since last successful run
+	// +optional
+    FailuresSinceSuccess int64
+
+	// The set of conditions for this object
+	// +optional
+	Conditions []CronJobCondition
+
+}
+```
+
+## v1beta1 changes
+All the new fields describedin the v1 section would be introduced in the v1beta1 API as well.
+
+
+## Validations
+Nothing additional from v1beta1
+
+## Plan for promoting CronJob to GA
+
+Dual controller (old and new coexist). Feature flag for new controller. alpha (disabled), beta(enabled, can be disabled), ga(enabled)
```

Comment:
Maybe this has been discussed, but I think that we don't really need Alpha for it.
We're going to use a standard pattern and reuse a bunch of existing logic for controller.

So while I agree it makes sense to not do in-place replacement, I would say that going directly to beta isn't that huge risk.
@liggitt - or would it be against some rules that I should be aware of?

Reply:
We discussed this in sig-apps ([notes](https://docs.google.com/document/d/1LZLBGW2wRDwAfdBNHJjFfk9CFoyZPcIYGWU7R1PQ3ng/edit#bookmark=id.7l4ut2t76aq4), [video](https://www.youtube.com/watch?v=R2iR5mxCyWA&list=PL69nYSiGNLP2LMq7vznITnpd2Fk1YIZF3&index=2&t=2m00s)).

If I understand correctly, the goal was to avoid rolling out a brand new controller implementation to all current consumers of the cronjob API simultaneously. Even though the current API is beta-level, it is very old and has attracted lots of use. Having a way to flip between the current implementation (which definitely has problems but is a known quantity) and the new implementation until the new implementation had gotten some soak time with actual users seemed beneficial.

Reply:
We already have a batch/v1beta1 and batch/v2alpha1 already for CronJob. 
So going to GA with 2 changes (API GA and new controller) and giving the user to choose the older implementation as a fallback is the goal here. 

Realized i had not reworded the bullet points. Rewording this paragraph.